Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+    {
+        modelBuilder.Entity<ShoppingCartModel>()
+            .ToTable("ShoppingCarts")
+            .OwnsMany(b => b.ProductItems, a =>
+            {
+                a.ToTable("ProductItems");
+                a.WithOwner().HasForeignKey("ShoppingCartId");
+                a.HasKey("ProductId", "ShoppingCartId");
+            });
+    }
+
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) =>
+        optionsBuilder
+            .UseInMemoryDatabase("ECommerceTest");
+}
+
+public class ShoppingCartRepository(ECommerceDBContext dbContext): IShoppingCartRepository
+{
+    public ValueTask<ShoppingCartModel?> Find(Guid id, CancellationToken ct) =>
+        dbContext.FindAsync<ShoppingCartModel>([id], ct);
+
+    public Task Store(Guid id, ShoppingCartModel? current, ShoppingCartEvent[] events, CancellationToken ct)
+    {
+        foreach (var @event in events)
+        {
+            switch (@event, current)
+            {
+                case (Opened(var clientId, var openedAt), null):
+                    dbContext.Add(new ShoppingCartModel
+                    {
+                        Id = id,
+                        ClientId = clientId,
+                        Status = ShoppingCartModel.ShoppingCartStatus.Pending,
+                        OpenedAt = openedAt
+                    });
+                    break;
+                case (ProductAdded(var (productId, quantity, price), var addedAt), not null):
+                    var toUpdate = current.ProductItems.SingleOrDefault(p => p.ProductId == productId);
+
+                    if (toUpdate == null)
+                    {
+                        current.ProductItems.Add(new ShoppingCartModel.PricedProductItem
+                        {
+                            ProductId = productId, Quantity = quantity, Price = price, AddedAt = addedAt,
+                        });
+                    }
+                    else
+                    {
+                        toUpdate.Quantity += quantity;
+                        toUpdate.UpdatedAt = addedAt;
+                    }
+
+                    break;
+                case (ProductRemoved(var (productId, quantity, price), var removedAt), not null):
+                    var existing = current.ProductItems.Single(p => p.ProductId == productId);
+
+                    if (existing.Quantity - quantity == 0)
+                    {
+                        current.ProductItems.Remove(existing);
+                    }
+                    else
+                    {
+                        existing.Quantity -= quantity;
+                        existing.UpdatedAt = removedAt;
+                    }
+
+                    break;
+                case (Confirmed (var clientId, var confirmedAt), not null):
+                    current.Status = ShoppingCartModel.ShoppingCartStatus.Confirmed;
+                    current.ClientId = clientId;
```

Comment:
whats the objective of this? why does it win over the opened one? why is the opened one optional?
unless its making an importtant point, I'd lose it?

Reply:
To show that client id may come later. Usually, user is forced to login upon shopping cart confirmation.